repair/repair.cc

```diff
@@ -2102,8 +2102,19 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
             rlogger.debug("repair[{}] Table {}.{} does not exist anymore", rid.uuid(), keyspace_name, table_name);
             continue;
         }
-        // FIXME: we need to wait for current tablet movement and disable future tablet movement
-        auto erm = t->get_effective_replication_map();
+        locator::effective_replication_map_ptr erm;
+        while (true) {
+            _repair_module->check_in_shutdown();
+            erm = t->get_effective_replication_map();
+            const locator::tablet_map& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
+            if (!tmap.has_transitions()) {
+                break;
+            }
+            rlogger.info("repair[{}] Table {}.{} has tablet transitions, waiting for topology to quiesce", rid.uuid(), keyspace_name, table_name);
+            erm = nullptr;
+            co_await _ss.local().await_topology_quiesced();
+            rlogger.info("repair[{}] Topology quiesced", rid.uuid());
```

Comment:
should await_topology_quiesced() return group0_guard? maybe not for this usage, but for others that might yield and want to read group0 state.

Reply:
There's no need, the caller can also obtain the guard. If topology is not quiesced then, he can retry. That's what await_topology_quiesced() would have to do internally anyway to return the guard.

Reply:
After the while (true) loop exits, what does it guarantee exactly?

Will holding a copy of the erm prevent any future topology operations / tablet migration? If so, how does it prevent the operations under the hood?

Reply:
It guarantees that erm doesn't have tablets in transition.

Holding to that erm prevents transitions from starting, because topology coordinator waits for old erms to be released when it executes global token metadata barrier.

Reply:
> It guarantees that erm doesn't have tablets in transition.

Only for the table in question which the erm belongs to, or for all tables?

Or even stronger, that no tablets (from any tables) in transitions and no topology changes cluster wide.  Since await_topology_quiesced waits for topology_state_machine to be not busy, it means the cluster is not in any of the states (tablet_draining, tablet_migration etc.).

```
future<> storage_service::await_topology_quiesced() {
    auto holder = _async_gate.hold();

    if (this_shard_id() != 0) {
        // group0 is only set on shard 0.
        co_await container().invoke_on(0, [&] (auto& ss) {
            return ss.await_topology_quiesced();
        });
        co_return;
    }

    group0_guard guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});

    while (_topology_state_machine._topology.is_busy()) {
        rtlogger.debug("await_topology_quiesced(): topology is busy");
        release_guard(std::move(guard));
        co_await _topology_state_machine.event.wait();
        guard = co_await _group0->client().start_operation(&_group0_as, raft_timeout{});
    }
}

bool topology::is_busy() const {
    return tstate.has_value();
}   

struct topology {
    enum class transition_state: uint16_t {
        join_group0,
        commit_cdc_generation,
        tablet_draining,
        write_both_read_old,
        write_both_read_new,
        tablet_migration,
        tablet_split_finalization,
        left_token_ring,
        rollback_to_normal,
    };

    std::optional<transition_state> tstate;

}
```

> Holding to that erm prevents transitions from starting, because topology coordinator waits for old erms to be released when it executes global token metadata barrier.

Including any tablet migration and adding/removing node as well?

Reply:
> Only for the table in question which the erm belongs to, or for all tables?

Only for the table, because that's what we check for.

Reply:
> Including any tablet migration and adding/removing node as well?

Yes.

Reply:
> Since await_topology_quiesced waits for topology_state_machine to be not busy, it means the cluster is not in any of the states (tablet_draining, tablet_migration etc.).

Yes, it waits for !busy, but this can change before we obtain the erm, so it's not guaranteed. What is guaranteed is what we check about the erm state.

Reply:
So given that why check `busy` at all? Isn't it enough to do:

```
while (auto erm = t->get_effective_replication_map(); erm.has_transitions()) {
    erm.reset();
    topology.even.wait();
}
```

Reply:
I think it's not that simple because topology is only valid on shard 0, so we cannot atomically check erm (on non-0 shard) and call wait(). We would be prone to missed notification if we check erm across shards, if topology becomes non-busy before we wait(). If we check erm on shard-0, then we can't use it on the original shard. We'd have to check the erm on both shard-0 and here.

Reply:
Will this work:

```
while (auto erm = t->get_effective_replication_map(); erm.has_transitions()) {
    erm.reset();
    invoken_on(0, [] () {
        if (t->get_effective_replication_map().has_transitions()) {
            topology.event.wait();
        }
   }
}
```

Reply:
No because "t" cannot be accessed across shards. We'd have to re-lookup it. 

Calling quiesce() is simpler.

Reply:
> No because "t" cannot be accessed across shards. We'd have to re-lookup it.
> 

Well, of course I did not mean the code to be 100% correct. erm does not have  `has_transitions` either.

> Calling quiesce() is simpler.

May be, but not much TBH. I just dislike the function since we still argue about what it actually guaranties and would gladly get rid of it.

Reply:
> May be, but not much TBH. I just dislike the function since we still argue about what it actually guaranties and would gladly get rid of it.

I can rename it to `await_not_busy()` to avoid confusion about semantics.

Reply:
It is better, but if this is the only change no need to delay the PR just for that.

Also @bhalevy here https://github.com/scylladb/scylladb/pull/18641#discussion_r1624511471 asked if we can wait for a specific tablet which I think my proposal would do.

Reply:
> > Also @bhalevy here [#18641 (comment)](https://github.com/scylladb/scylladb/pull/18641#discussion_r1624511471) asked if we can wait for a specific tablet
> 
> Specific table, not tablet. That latter is too complicated at the moment. I'm not sure about the former.

Yes, I meant table.

Reply:
> Also @bhalevy here https://github.com/scylladb/scylladb/pull/18641#discussion_r1624511471 asked if we can wait for a specific tablet which I think my proposal would do.

Yes, but I don't think we need to optimize for this. I'd rather invest in implementing to proper solution, which is to schedule repair from tablet scheduler.